spylib/utils/session_token.py

```diff
@@ -0,0 +1,106 @@
+from __future__ import annotations
+
+from typing import Any, Dict, List, Optional
+from urllib.parse import urlparse
+
+from jwt import decode
+from pydantic import root_validator
+from pydantic.main import BaseModel
+from pydantic.networks import HttpUrl
+
+from .domain import store_domain
+
+
+class ValidationError(Exception):
+    pass
+
+
+class InvalidIssuerError(ValidationError):
+    pass
+
+
+class MismatchedHostError(ValidationError):
+    pass
+
+
+class TokenAuthenticationError(ValidationError):
+    pass
+
+
+class SessionToken(BaseModel):
+    """
+    Session tokens are derived from the authrization header from Shopify.
+    This performs the set of validations as defined by shopify
+    https://shopify.dev/apps/auth/session-tokens/authenticate-an-embedded-app-using-session-tokens#obtain-session-details-manually
+    """
+
+    iss: HttpUrl
+    dest: HttpUrl
+    aud: Optional[str]
+    sub: int
+    exp: Optional[float]
+    nbf: Optional[float]
+    iat: Optional[float]
+    jti: int
+    sid: str
+
+    @root_validator()
+    def equal_iss_and_dest(cls, values: Dict[str, Any]):
+        domain = cls.__url_to_base(values.get('iss'))
+        try:
+            store_domain(domain)
+        except ValueError as e:
+            raise InvalidIssuerError(f"The domain {domain} is not a valid issuer.") from e
+
+        iss = cls.__url_to_base(values.get('iss'))
+        dest = cls.__url_to_base(values.get('dest'))
+        if iss != dest:
+            raise MismatchedHostError(f'The issuer {iss} does not match the destination {dest}')
+
+        return values
+
+    @classmethod
+    @property
+    def algorithm(cls) -> str:
+        return "HS256"
+
+    @classmethod
+    @property
+    def prefix(cls) -> str:
+        return "Bearer "
+
+    @classmethod
+    @property
+    def required_fields(cls) -> List[str]:
+        return ["iss", "dest", "sub", "jti", "sid"]
+
+    @classmethod
+    def from_header(
+        cls,
+        authorization_header: str,
+        api_key: str,
+        secret: str,
+    ) -> SessionToken:
+
+        # Take the authorization headers and unload them
+        if not authorization_header.startswith(str(cls.prefix)):
+            raise TokenAuthenticationError(
+                "The authorization header does not contain a Bearer token."
+            )
+
+        token = authorization_header[len(str(cls.prefix)) :]
```

Comment:
What you need to [know about regex](https://www.python-course.eu/python3_re.php) and a [handy regex checker](https://regex101.com/)

```suggestion
        if result := re.search(r'Bearer (.*)', authorization_header):
            token = result.group(1)
        else:
            raise TokenAuthenticationError(
                "The authorization header does not contain a Bearer token."
            )
```

Reply:
Is this just for performance? As if we use the raw string 'Bearer ' I feel like `startswith()` is faster/easier to read:

```python
if not authorization_header.startswith('Bearer '):
  raise TokenAuthenticationError(
    "The authorization header does not contain a Bearer token."
  )
```

Reply:
I like regexp but I might be the only one ... :sweat_smile: 
It's fine either way